Core/include/Acts/Utilities/KDTree.hpp

```diff
@@ -318,369 +310,249 @@ class KDTree {
-
-    /// @brief Destroy this node in the k-d tree, and indirectly all of its
-    /// children.
-    virtual ~KDTreeAbstractNode(void) = default;
-
-    /// @brief Perform a range search in the k-d tree, mapping the key-value
-    /// pairs to a side-effecting function.
-    ///
-    /// This is the most powerful range search method we have, assuming that we
-    /// can use arbitrary side effects, which we can. All other range search
-    /// methods are implemented in terms of this particular function.
-    ///
-    /// @param r The range to search for.
-    /// @param f The mapping function to apply to matching elements.
-    virtual void rangeSearchMapDiscard(
-        const range_t &r,
-        std::function<void(const coordinate_t &, const Type &)> f) const = 0;
-
-    /// @brief Debugging method that gives a textual representation of a k-d
-    /// tree.
-    ///
-    /// It is probably best not to use this method in real code. It is designed
-    /// for debugging purposes only.
-    ///
-    /// @param d The depth of this node (for indentation).
-    virtual std::string toString(std::size_t d) const = 0;
+          m_range(boundingBox(m_begin_it, m_end_it)) {
+      if (m_type == NodeType::INTERNAL) {
+        // This constant determines the maximum number of elements where we
+        // still
+        // calculate the exact median of the values for the purposes of
+        // splitting. In general, the closer the pivot value is to the true
+        // median, the more balanced the tree will be. However, calculating the
+        // median exactly is an O(n log n) operation, while approximating it is
+        // an O(1) time.
+        constexpr std::size_t max_exact_median = 128;
+
+        // The following set of operations is designed to find the variance (of
+        // normalized values) of the values in each of the dimensions. We do
+        // this because the optimal way of constructing a k-d tree is to select
+        // the pivot dimension as the dimension with the highest variance. To
+        // this end, we first start by calculating the sum of all the values.
+        std::array<Scalar, Dims> sum_v{};
+
+        sum_v.fill(0.0);
 
-    /// @brief Determine the number of elements managed by this node.
-    ///
-    /// Conveniently, this number is always equal to the distance between the
-    /// begin iterator and the end iterator, so we can simply delegate to the
-    /// relevant standard library method.
-    ///
-    /// @return The number of elements below this node.
-    std::size_t size(void) const { return std::distance(m_begin_it, m_end_it); }
-
-    /// @brief The axis-aligned bounding box containing all elements in this
-    /// node.
-    ///
-    /// @return The minimal axis-aligned bounding box that contains all the
-    /// elements under this node.
-    const range_t &range(void) const { return m_range; }
+        for (iterator_t i = this->m_begin_it; i != this->m_end_it; ++i) {
```

Comment:
We usually omit `this`, but you were using this in the previous version of the code already, so it's fine.